### PR TITLE
[ROCm][Quantization] add fp8xfp8 attn support for rocm_aiter_unified_attn

### PR DESCRIPTION
**Pull request by divakar-amd**
As a follow-up of https://github.com/vllm-project/vllm/pull/35316, 
This PR adds support for fp8xfp8 attention for `rocm_aiter_unified_attn` 
For fp8xfp8 computation, the aiter triton kernel expects Q to be in fp8 as well, otherwise it descales it to fp32. [(LINK)](https://github.com/ROCm/aiter/blob/3a217fc9885a5d5f799c2f2aacecabf9f5e99614/aiter/ops/triton/_triton_kernels/attention/unified_attention.py#L270-L276). However, Q scales are not supported in the kernel ([LINK](https://github.com/ROCm/aiter/blob/a94287503368d1c64f6d3be94b5704501d8c1774/aiter/ops/triton/attention/unified_attention.py#L127))
To handle this, we absorb `q_scale` & `k_scale` into the `softmax_scale`, and absorb `v_scale` into the `output_scale` (if fused quantized output is used) or apply it post-attention.

## Testing on gpt-oss-120b (On top of https://github.com/vllm-project/vllm/pull/35316)

<details>
  <summary>**Quark quantising script for gpt-oss-120b**</summary>

  > #/bin/bash
exclude_layers="\*lm_head* \*router*"
python3 quantize_quark.py \
  --model_dir /data/models/openai/gpt-oss-120b \
  --quant_scheme mxfp4_fp8 \
  --kv_cache_dtype fp8 \
  --attention_dtype fp8 \
  --exclude_layers $exclude_layers \
  --num_calib_data 512 \
  --output_dir /data/models/gpt-oss-120b-w-mxfp4-a-fp8-kv-fp8-fp8attn-no_lmhead_router \
  --model_export hf_format \
  --multi_gpu
</details>

**Server Cmd**
> VLLM_ROCM_USE_AITER=1 VLLM_ROCM_USE_AITER_UNIFIED_ATTENTION=1 vllm serve /data/models/gpt-oss-120b-w-mxfp4-a-fp8-kv-fp8-fp8attn-no_lmhead_router --kv_cache_dtype="fp8"

**Client Cmd**
> curl http://localhost:8000/v1/chat/completions -H "Content-Type: application/json" -d '{
"model":"/data/models/gpt-oss-120b-w-mxfp4-a-fp8-kv-fp8-fp8attn-no_lmhead_router",
"messages": [
{"role": "system", "content": "You are a helpful assistant."},
{"role": "user", "content": "Explain briefly the process of photosynthesis."}
],
"max_tokens": 200,
"temperature": 0.7
}'

**Output**
> {"id":"chatcmpl-9d28b156d199bda2","object":"chat.completion","created":1773346788,"model":"/data/models/gpt-oss-120b-w-mxfp4-a-fp8-kv-fp8-fp8attn-no_lmhead_router","choices":[{"index":0,"message":{"role":"assistant","content":"**Photosynthesis – a quick overview**\n\n1. **Where it happens**  \n   - In the chloroplasts of plant (and some algal) cells, mainly in the thylakoid $embranes (light‑dependent part) and the stroma (light‑independent part).\n\n2. **Main stages**  \n   - **Light‑dependent reactions** – occur in the thylakoid membranes. Light energy excites chlorophyll, driving the “water‑splitting” (photolysis) reaction that releases O₂, generates ATP and NADPH.  \n   - **Calvin‑Benson (light‑independent) cycle** –uses","refusal":null,"annotations":null,"audio":null,"function_call":null,"tool_calls":[],"reasoning":"User asks: \"Explain briefly the process of photosynthesis.\" So we need to give a concise explanation: what it is, where it occurs, overall equation, light-dependent and light-independent reactions, chlorophyll, etc. Probably a short paragraph or bullet points. Provide a clear overview."},"logprobs":null,"finish_reason":"length","stop_reason":null,"token_ids":null}],"service_tier":null,"system_fingerprint":null,"usage": {"prompt_tokens":85,"total_tokens":285,"completion_tokens":200,"prompt_tokens_details":null},"prompt_logprobs":null,"prompt_token_ids":null,"kv_transfer_params":null}

**LM-eval**
> lm_eval --model local-completions --model_args "base_url=http://localhost:8000/v1/completions,pretrained=/data/models/gpt-oss-120b-w-mxfp4-a-fp8-kv-fp8-fp8attn-no_lmhead_router,tensor_parallel_size=1,add_bos_token=true,trust_remote_code=true,max_gen_toks=2048" --tasks gsm8k --num_fewshot 5 --batch_size 64 --limit 250 --apply_chat_template --seed 42

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.880|±  |0.0206|
|     |       |strict-match    |     5|exact_match|↑  |0.052|±  |0.0141|

### CODE REVIEW

File: vllm/v1/attention/backends/rocm_aiter_unified_attn.py [hunk 1]
```diff
@@ -190,12 +191,20 @@ def forward(
...
-            assert layer._q_scale_float == 1.0, (
-                "A non 1.0 q_scale is not currently supported."
-            )
+            # When Q is FP8, triton kernel skips K/V dequant (for fp8xfp8 matmul).
+            # Compensate by absorbing q_scale and k_scale into softmax_scale, and
+            # v_scale into output_scale (or post-multiplying if no fusion).
+            if query.dtype == self.fp8_dtype:
+                softmax_scale = self.scale * layer._q_scale_float * layer._k_scale_float
+                if output_scale is not None:
+                    output_scale = output_scale / layer._v_scale_float
```
[the remark this answers is not in the document]
Comment (divakar-amd) [reply vllm/v1/attention/backends/rocm_aiter_unified_attn.py:205]: Nope! Aiter inverts the output scale. Check this out - https://github.com/ROCm/aiter/blob/3a217fc9885a5d5f799c2f2aacecabf9f5e99614/aiter/ops/triton/attention/unified_attention.py#L203

File: vllm/v1/attention/backends/rocm_aiter_unified_attn.py [hunk 2]
```diff
@@ -190,12 +191,20 @@ def forward(
 
         key_cache, value_cache = kv_cache.unbind(0)
 
+        softmax_scale = self.scale
```
Comment (BowenBao) [inline vllm/v1/attention/backends/rocm_aiter_unified_attn.py:194]: what is this `self.scale`?

File: vllm/v1/attention/backends/rocm_aiter_unified_attn.py [hunk 3]
```diff
@@ -190,12 +191,20 @@ def forward(
...
-            )
+            # When Q is FP8, triton kernel skips K/V dequant (for fp8xfp8 matmul).
+            # Compensate by absorbing q_scale and k_scale into softmax_scale, and
+            # v_scale into output_scale (or post-multiplying if no fusion).
+            if query.dtype == self.fp8_dtype:
+                softmax_scale = self.scale * layer._q_scale_float * layer._k_scale_float
+                if output_scale is not None:
+                    output_scale = output_scale / layer._v_scale_float
+                else:
+                    fp8_post_attn_v_rescale = True
```
Comment (BowenBao) [inline vllm/v1/attention/backends/rocm_aiter_unified_attn.py:207]: could we make `output_scale = 1 / layer._v_scale_float` in this case?

File: vllm/v1/attention/backends/rocm_aiter_unified_attn.py [hunk 4]
```diff
@@ -125,6 +125,7 @@ def __init__(
         from aiter.ops.triton.unified_attention import unified_attention
 
         self.unified_attention = unified_attention
+        self.supports_quant_query_input = True
```
Comment (BowenBao) [inline vllm/v1/attention/backends/rocm_aiter_unified_attn.py:128]: is this used somewhere else?

**Review by BowenBao — COMMENTED** [review]
Discussed offline, thanks for explaining the absorption trick. It'd be much helpful if 'kernel_unified_attention_2d' and 'unified_attention' had some comments explaining what their definition of these scales are.

I have a few other questions but overall looks good.

File: vllm/v1/attention/backends/rocm_aiter_unified_attn.py (hunk 4 above)
Comment (divakar-amd) [reply vllm/v1/attention/backends/rocm_aiter_unified_attn.py:128]: Yes, it's required for the following check (to create fp8 query): [LINK](https://github.com/vllm-project/vllm/blob/4508532fbd299cff81ecb6f1ccea2e2d0f56d329/vllm/model_executor/layers/attention/attention.py#L363)

File: vllm/v1/attention/backends/rocm_aiter_unified_attn.py (hunk 3 above)
Comment (divakar-amd) [reply vllm/v1/attention/backends/rocm_aiter_unified_attn.py:207]: I believe not, since value of `output_scale` would be non-zero when fusion path is used (i.e. compiler's pattern matching to directly produce fp8 attn output for the next layer)

Comment (BowenBao) [reply vllm/v1/attention/backends/rocm_aiter_unified_attn.py:207]: I'm referring to the else branch where `output_scale is None`.

File: vllm/v1/attention/backends/rocm_aiter_unified_attn.py (hunk 2 above)
Comment (divakar-amd) [reply vllm/v1/attention/backends/rocm_aiter_unified_attn.py:194]: It is the attention scaling factor:  `1 / sqrt(head_size)`

File: vllm/v1/attention/backends/rocm_aiter_unified_attn.py (hunk 3 above)
Comment (divakar-amd) [reply vllm/v1/attention/backends/rocm_aiter_unified_attn.py:207]: This is a great suggestion. The only concern is that there is an [FP8 clamping](https://github.com/ROCm/aiter/blob/f40d508ebbe90d74b843a4207dce062a4066b4cf/aiter/ops/triton/_triton_kernels/attention/unified_attention.py#L365) occuring inside the kernel even though the expected output is in fp16/bf16. Let's stick to the current implementation for now and revisit this later, once the Aiter kernel has been updated.

**Review by BowenBao — APPROVED** [review]

**Review by tjtanaa — COMMENTED** [review]
LGTM, one last thing before merging, is it validated with AITER_BRANCH="v0.1.10.post2" ? the aiter specified in the `Dockerfile.rocm_base` ?

**Review by tjtanaa — APPROVED** [review]
LGTM
